pkg/proc/target_exec.go

```diff
@@ -98,31 +105,56 @@ func (grp *TargetGroup) Continue() error {
 			// Issue #2078.
 			// Errors are ignored because depending on why ContinueOnce failed this
 			// might very well not work.
-			if valid, _ := dbp.Valid(); valid {
-				if trapthread != nil {
-					_ = dbp.SwitchThread(trapthread.ThreadID())
-				} else if curth := dbp.CurrentThread(); curth != nil {
-					dbp.selectedGoroutine, _ = GetG(curth)
-				}
-			}
+			_ = grp.fixCurrentThreads(traptgt, trapthread)
 			if pe, ok := contOnceErr.(ErrProcessExited); ok {
-				dbp.exitStatus = pe.Status
+				traptgt.exitStatus = pe.Status
 			}
 			return contOnceErr
 		}
-		if dbp.StopReason == StopLaunched {
-			dbp.ClearSteppingBreakpoints()
+		if stopReason == StopLaunched {
+			it := ValidTargets{Group: grp}
+			for it.Next() {
+				it.Target.ClearSteppingBreakpoints()
+			}
 		}
 
-		callInjectionDone, callErr := callInjectionProtocol(dbp, threads)
-		hcbpErr := dbp.handleHardcodedBreakpoints(trapthread, threads)
+		var callInjectionDone bool
+		var callErr error
+		var hcbpErr error
+		it = ValidTargets{Group: grp}
```

Comment:
I don't like using closures for looping because break and continue don't work with them and return doesn't do what you want it to do.

It also doesn't simplify the code much, for example this:

```
it := ValidTargets{Group: grp}
for it.Next() {
	for _, bp := range it.Breakpoints().M {
		if bp.LogicalID() == lbp.LogicalID {
			n++
			err := it.ClearBreakpoint(bp.Addr)
			if err != nil {
				errs = append(errs, err)
			}
		}
	}
}
```

becomes

```
grp.ForEachValidTarget(func (t *proc.Target) error {
	for _, bp := range t.Breakpoints().M {
		if bp.LogicalID() == lbp.LogicalID {
			n++
			err := t.ClearBreakpoint(bp.Addr)
			if err != nil {
				errs = append(errs, err)
			}
		}
	}
	return nil
})
```

which is the exact same number of lines.

Reply:
Valid points.